### PR TITLE
STM32duino support

### PR DESCRIPTION
**Pull request by ubis**
Finally found some time to make PR.. Anyways, a package named "framework-stm32duino" should be created from [https://github.com/rogerclarkmelbourne/Arduino_STM32](https://github.com/rogerclarkmelbourne/Arduino_STM32), I didn't do any changes to it so it should be kept up-to date and work out of the box.

However, I did noticed that builds are slightly bigger than compiled with Arduino IDE and I'm not sure what's causing this, I might investigate this a bit later, as well as adding another boards.

### CODE REVIEW

**Comment by ivankravets** [comment]
Thanks a lot for great job!!! I have a few questions:

1. You propose to add a few new boards `duino_***`. Do these boards really exist? If no, I propose to use `generic_***` prefix.
2. `"frameworks": ["stm32duino"]` should be replaced with ` "frameworks": ["arduino"]`, because STM32Duino is Wiring/Arduino based framework. Developers will have fewer questions. What is more, we use in code `#include <Arduino.h>` but not `#include <STM32Duino.h>`. In addition, all libraries that compatible with Arduino framework, will be automatically compatible with STM23Duino

Also, I'll make some code review after new PR.

**Comment by ivankravets** [comment]
I've just checked original source code and foud all boards https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/master/STM32F1/boards.txt

* mapleMini
* microduino32_flash
* genericSTM32F103C
* ...

These are right board IDs. I propose to use them.

The board presets:
* https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/master/STM32F1/boards.txt
* https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/master/STM32F3/boards.txt
* https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/master/STM32F4/boards.txt

Would be thankful if you will use the same information.

**Comment by ubis** [comment]
I used that _duino__ prefix just to able see which boards came from stm32duino framework, since there will be some mess(it has many boards), but I can rename them, not a problem. 

As for STM32F3/F4, I didn't added support for them yet, but I will, later with new PR's.

**Comment by ivankravets** [comment]
We don't link each board to framework. A single board could be compatible with many frameworks. I hope you understand me. Otherwise, we will have 1000 supported boards with identical hardware. This is not good.

**Comment by ubis** [comment]
Yeah, that seems logical ;)

**Comment by ivankravets** [comment]
Please use this manifest https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/master/STM32F1/boards.txt and its board IDs

**Comment by ivankravets** [comment]
```
 "url": "http://www.stm32duino.com", 
 "vendor": "STM32duino"
```
Please set it to `vendor = Generic" and "url" to `http://www.st.com/content/st_com/en/products/microcontrollers/stm32-32-bit-arm-cortex-mcus/stm32f1-series.html?querycriteria=productId=SS1031`. See full map with valid links http://www.st.com/content/st_com/en/products/microcontrollers/stm32-32-bit-arm-cortex-mcus.html?querycriteria=productId=SC1169

P.S: I don't cavil to your work. Just want to keep all right.

**Comment by ivankravets** [comment]
What is a difference between `build.hwids` and `upload.usbid`? PlatformIO uses `build.hwids` to find proper upload port. I propose to merge `upload.usbid` with `build.hwids`, where the first pair in `build.hwids` will be used for uploading (if you need it).

It will allow to detect devices under multiple HWIDs.

**Comment by ubis** [comment]
Mainly, stm32duino framework uses these upload.usbid's for flashing via DFU protocol, but it has build.hwid's in board configs aswell, probably for developing usb devices. Will merge it to hwid's then.

File: builder/main.py [hunk 1]
```diff
@@ -79,6 +77,14 @@
     PROGSUFFIX=".elf"
 )
 
+if not "stm32duino" in env.subst("$PIOFRAMEWORK"):
```
Comment (ivankravets) [inline builder/main.py]: I propose to move this logic to `arduino.py` file. You can remove values from env variable. For example,
```python
for item in ("-nostartfiles", "-nostdlib"):
    if item in env['LINKFLAGS']:
        env['LINKFLAGS'].remove(item)
```

File: builder/main.py [hunk 2]
```diff
@@ -140,6 +146,32 @@
         UPLOADCMD='$UPLOADER $UPLOADERFLAGS'
     )
 
+if "stm32duino" in env.subst("$PIOFRAMEWORK"):
```
Comment (ivankravets) [inline builder/main.py]: `if "arduino"`...

File: builder/main.py [hunk 3]
```diff
@@ -174,6 +206,12 @@
         [env.VerboseAction(env.AutodetectUploadPort,
                            "Looking for upload disk..."),
          env.VerboseAction(env.UploadToDisk, "Uploading $SOURCE")])
+elif "stm32duino" in env.subst("$PIOFRAMEWORK"):
```
Comment (ivankravets) [inline builder/main.py]: `elif arduino`

File: builder/main.py [hunk 4]
```diff
@@ -140,6 +146,32 @@
         UPLOADCMD='$UPLOADER $UPLOADERFLAGS'
     )
 
+if "stm32duino" in env.subst("$PIOFRAMEWORK"):
+    uploadProtocol = ""
+    uploadParams = ""
+    from sys import platform as _platform
```
Comment (ivankravets) [inline builder/main.py]: Please use append on the top of file
```python
from platformio import util

...

# the use 
if "linux" in util.get_systype():
   # code..

**Comment by ivankravets** [comment]
The build script for framework should have name `builder/frameworks/arduino.py` and package will have name `framework-arduinoststm32`.

**Comment by ubis** [comment]
Yeah, some parts could be written in cleaner/better way, but since i'm programming in Python for the first time, I don't know much about this lang, yet.

**Comment by ivankravets** [comment]
> but since i'm programming in Python for the first time

First time? Your code is cool for the first time! Did you have experience with SCons?

**Comment by ubis** [comment]
Nope, I don't know anything about it.
I just try to keep it up with lang's syntax/coding style.

**Comment by ivankravets** [comment]
I'm joking! Excellent work! :)

**Comment by ivankravets** [comment]
Could we add all boards from this list? https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/master/STM32F1/boards.txt

File: builder/frameworks/arduino.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+from os import walk
+from os.path import isdir, isfile, join
+
+from SCons.Script import DefaultEnvironment
+
+env = DefaultEnvironment()
+platform = env.PioPlatform()
+
+if "stm32f103" in env.BoardConfig().get("build.mcu", ""):
+    FRAMEWORK_DIR = join(platform.get_package_dir("framework-stm32duino"), "STM32F1")
```
Comment (ivankravets) [inline builder/frameworks/arduino.py]: Please replace text in the all files from `framework-stm32duino` to `framework-arduinoststm32`

**Comment by ubis** [comment]
Thanks ;) also, I will keep it supported from my side(by adding other boards/flashing protocols in the future).

Yeah, sure, I will add them.

**Comment by ivankravets** [comment]
Now, please fork original repo https://github.com/rogerclarkmelbourne/Arduino_STM32

Add in the roo this file `package.json`
```json
{
    "description":"Arduino Wiring-based Framework (ST STM32 Core)",
    "name":"framework-arduinoststm32",
    "system":"all",
    "url":"https://github.com/rogerclarkmelbourne/Arduino_STM32",
    "version":"1.0.0"
}
```

You will maintain this package. If it will work good, we make PR to main repo with package manifest.

**Comment by ivankravets** [comment]
When you forked repo and added `package.json` manifest, please make first release `v1.0.0`.

Now, need to replace [packageRepositories](https://github.com/platformio/platform-ststm32/blob/develop/platform.json#L17) with the next content:
```
  "packageRepositories": [
    "https://dl.bintray.com/platformio/dl-packages/manifest.json",
    "https://sourceforge.net/projects/platformio-storage/files/packages/manifest.json/download",
    "http://dl.platformio.org/packages/manifest.json",
    {
      "framework-arduinoststm32": [
        {
          "url": "GitHub url to ZIP or TAZ.GZ archive",
          "sha1": "set here sha1 checksum",
          "version": "1.0.0",
          "system": "*"
        }
      ]
  ],
```

**Comment by ubis** [comment]
I've been adding these boards for the past 2 hours, and I can say I can't add all of them for now.
Some of them uses different compiler flags and I got tired searching for them already. Due some flags, some options in builder/frameworks/arduino.py might be changed.

I'll add them later, with another PR. And, as far as I know, the generic ones i've added are the mostly used within STM32duino.

**Comment by ivankravets** [comment]
What I can say? Thank you! :)

I'll try to test it now and contact you! Looks like I have supported board but I'm not sure.

**Comment by ubis** [comment]
Well, what I can say, I would add support for other boards aswell, if I had them. My generic STM32F103RB works. Other boards compiles aswell, but I can't test them. Should work though.

**Comment by ivankravets** [comment]
What is number of LED pin for `maple` and `genericSTM32F103RB`?

**Comment by ubis** [comment]
for generic its PC13, for maple i'm not sure PA5 maybe..?

**Comment by ivankravets** [comment]
Looks like P13 too 
http://www.stm32duino.com/viewtopic.php?t=58

**Comment by ivankravets** [comment]
I've just added a few examples https://github.com/platformio/platform-ststm32/tree/develop/examples

They compile good on my machine.

**Comment by ivankravets** [comment]
Could you write something here? http://www.stm32duino.com/viewtopic.php?f=41&t=728&p=14988&hilit=platformio#p14988

Need to test it before release. How to try?

1. Install [PlatformIO IDE](http://platformio.org/platformio-ide)
2. Open PlatformIO IDE Terminal and run this (install development version of ST STM32):
```
platformio platform install https://github.com/platformio/platform-ststm32.git
```
3. Initialize project with just added boards.

**Comment by ubis** [comment]
I wrote there. I hope they will try it.

**Comment by ivankravets** [comment]
Thanks!!! Keep me in touch!

Added new examples for CI too https://travis-ci.org/platformio/platform-ststm32/builds/158980621

**Comment by ubis** [comment]
I'm not sure if I should post here or create a new issue.
Seems like there is a possibly board duplicate: genericSTM32F103C8 and bluepill_f103c8.
Some users in STM32duino forum want to use bluepill with PlatformIO. but adding arduino option to config won't do the job. Build core, variants are different. As well as ldscript and uploading options.

I think, these options could be changed inside builder/frameworks/arduino.py if that board is used.
What do you think?

**Comment by ivankravets** [comment]
@ubis please open new issue and copy your comment. Also, please provide me 2 `board.json`, need to examine in details this situation. Thanks

**Pull request merged**